var/spack/repos/builtin/packages/upcxx/package.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+import llnl.util.tty as tty
+import os
+
+
+class Upcxx(Package):
+    """UPC++ is a C++ library that supports Partitioned Global Address Space
+    (PGAS) programming, and is designed to interoperate smoothly and
+    efficiently with MPI, OpenMP, CUDA and AMTs. It leverages GASNet-EX to
+    deliver low-overhead, fine-grained communication, including Remote Memory
+    Access (RMA) and Remote Procedure Call (RPC)."""
+
+    homepage = "https://upcxx.lbl.gov"
+
+    version('2019.3.2', '844722cb0e8c0bc649017fce86469457')
+
+    variant('cuda', default=False, 
+            description='Builds a CUDA-enabled version of UPC++')
+    depends_on('cuda', when='+cuda')
+
+    def url_for_version(self, version):
+        if version > Version('2019.3.2'):
+            url = "https://bitbucket.org/berkeleylab/upcxx/downloads/upcxx-{0}.tar.gz"
+        else:
+            url = "https://bitbucket.org/berkeleylab/upcxx/downloads/upcxx-{0}-offline.tar.gz"
+        return url.format(version)
+
+    def setup_environment(self, spack_env, run_env):
+        if 'platform=cray' in self.spec:
+            if (os.environ.get('CROSS') is None):
+                srunbin = which('srun')
+                if srunbin is None:
+                    aprunbin = which('aprun')
+                    if aprunbin is None:
+                        tty.error("Detected a Cray system. CROSS has not been"
+                                  " set, but cannot find srun nor aprun."
+                                  " Aborting now.")
+                    else:
+                        spack_env.set('CROSS', 'cray-aries-alps')
+                        tty.warn("Detected a Cray system. CROSS has not been"
+                                 " set, aprun has been found so assuming"
+                                 " CROSS=cray-aries-alps")
+                else:
+                    spack_env.set('CROSS', 'cray-aries-slurm')
+                    tty.warn("Detected a Cray system. CROSS has not been"
+                             " set, srun has been found so assuming"
+                             " CROSS=cray-aries-slurm")
+
+                spack_env.set('GASNET_CONFIGURE_ARGS', '--enable-mpi=probe')
```

Comment:
@bonachea @mjacquel: One last request and then this LGTM.  We are looking at ways we could better detect things like the resource manager on the system, but we don't have better ways than what you're doing here.

However, we're also trying to make all our builds cacheable, and the fact that this stuff is environment-dependent and does *not* expose it in the package metadata in any way will prevent that.

Can you please expose the things you want to detect as variants?  Essentially that would mean factoring out the logic here as methods and doing something like this:

```python

def detect_scheduler():
    # ... do something to return `'alps'`, `'slurm'`, or `'none'`

class Upcxx(Package):

    variant('scheduler', values=('slurm', 'alps', 'none'), default=detect_scheduler(), 
            description="resource manager to use")


    def setup_environment(self, spack_env, run_env):
        if "scheduler=slurm" in self.spec:
            # set env vars, etc.
        else:
            # ...
```

If you do this then the variant's default will still depend on the environment on the platform where you run `spack`, but it will be exposed in the package metadata, which means that we can safely cache this build configuration.  If the build can differ depending on the environment, we can't do that.

Does that make sense?

Reply:
@tgamblin  - we have an underlying assumption here that there is only one "right" answer to the question of resource manager for a given system. ie we've never encountered a Cray XC where the policies permit users to freely choose between SLURM and ALPS, and can't imagine that ever being deployed in practice. This means a given site will only ever install the one correct "variant", and it would never make sense to install the other one on the same system, because it would be unable to run jobs. 

So the only value I see to the "variant" you propose is that on a (theoretical) system with SLURM binaries installed but where the correct answer is actually ALPS, one would use the variant to override the default and select ALPS, instead of the `CROSS` envvar.  Does that match the motivation for your request? (Note we know of no site meeting the criteria above)

It's tangentially worth noting that the `CROSS` input to UPC++ install may in the future encapsulate additional dimensions of freedom. The only cross-compiled platform supported in the _current_ release of UPC++ is the Cray XC (leaving only the SLURM/ALPS dimension, which cannot be auto-detected). However GASNet supports cross-compilation for other platforms and if UPC++ ever grows support for others then we may need to revisit this design at that time.

Reply:
Also worth noting the SLURM vs ALPS distinction is only meaningful to the UPC++ package on Cray XC systems, and has no effect elsewhere. On Cray XC systems the 'none' value is unacceptable and will lead to an install error.

Reply:
@bonachea:

> This means a given site will only ever install the one correct "variant", and it would never make sense to install the other one on the same system, because it would be unable to run jobs.

This isn't so much to give the user a choice as to differentiate between a *Spack* binary built for system A and one built for system B.  Sites may share binary caches (esp. under ECP) and if you do not add the variant, the hashes for the packages built at two different sites will be the same.  So we wouldn't be able to, e.g., provide a public binary mirror of builds for Spack.  We're thinking bigger than a single site.

So, I don't expect a user to `spack install upcxx scheduler=slurm` on an `alps` system, or vice versa.  But I expect the binaries produced by `spack buildcache create upcxx` to have different hashes on those two systems. Note that the logic I recommended moving out just controls the default value of the variant, so the user doesn't have to care.  `spack install upcxx` on a `slurm` machine will concretize to something with `scheduler=slurm`, and on an `alps` system it'll concretize to something with `scheduler=alps`.  The point is that we'll have the variant so we can pull the right binary from the buildcache and ultimately enable the user not to have to build at all.  Without the variant, these two different builds will have the same hash.

RE:
> Also worth noting the SLURM vs ALPS distinction is only meaningful to the UPC++ package on Cray XC systems, and has no effect elsewhere. On Cray XC systems the 'none' value is unacceptable and will lead to an install error.

This should be fine -- you can default to `'none'` when not on cray, and add:

`conflicts('scheduler=none', when='platform=cray', 
         msg='None is unacceptable on Cray.')`

 



This way, if users at either site

Reply:
RE:

> Also worth noting the SLURM vs ALPS distinction is only meaningful to the UPC++ package on Cray XC systems, and has no effect elsewhere. On Cray XC systems the 'none' value is unacceptable and will lead to an install error.

We're hoping to allow variants to have a `when` clause soon, so you could make the variant just not exist on other platforms, but at the moment it has to... but none shouldn't hurt anything

Reply:
@tgamblin : Thanks for the clarifications.

@mjacquel will look into deploying your suggested changes tonight.

Reply:
Thanks!

Reply:
Alright, I pushed changes which I think address your request @tgamblin. However, I messed up and apparently pushed a conflicting default config.yml (which I had to modify on NERSC Cori because I didn't have write access on the default stage directory).